pkg/controller/job/indexed_job_utils.go

```diff
@@ -0,0 +1,203 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package job
+
+import (
+	"fmt"
+	"math"
+	"sort"
+	"strconv"
+	"strings"
+
+	batch "k8s.io/api/batch/v1"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/kubernetes/pkg/controller"
+)
+
+const (
+	completionIndexEnvName = "JOB_COMPLETION_INDEX"
+	unknownCompletionIndex = -1
+)
+
+// calculateCompletedIndexesStr returns a string representation of the list
+// of completed indexes in compressed format.
+func calculateCompletedIndexesStr(pods []*v1.Pod) string {
+	sort.Sort(byCompletionIndex(pods))
+	var result strings.Builder
+	var lastSucceeded int
+	firstSucceeded := math.MinInt32
+	for _, p := range pods {
+		ix := getCompletionIndex(p.Annotations)
+		if ix == unknownCompletionIndex {
+			continue
+		}
+		if p.Status.Phase == v1.PodSucceeded {
+			if firstSucceeded == math.MinInt32 {
+				firstSucceeded = ix
+			} else if ix > lastSucceeded+1 {
+				addSingleOrRangeStr(&result, firstSucceeded, lastSucceeded)
+				firstSucceeded = ix
+			}
+			lastSucceeded = ix
+		}
+	}
+	if firstSucceeded != math.MinInt32 {
+		addSingleOrRangeStr(&result, firstSucceeded, lastSucceeded)
+	}
+	return result.String()
+}
+
+func addSingleOrRangeStr(builder *strings.Builder, first, last int) {
+	if builder.Len() > 0 {
+		builder.WriteRune(',')
+	}
+	builder.WriteString(strconv.Itoa(first))
+	if last > first {
+		if last == first+1 {
+			builder.WriteRune(',')
+		} else {
+			builder.WriteRune('-')
+		}
+		builder.WriteString(strconv.Itoa(last))
+	}
+}
+
+func appendDuplicatedIndexPodsForRemoval(rm, left, pods []*v1.Pod) ([]*v1.Pod, []*v1.Pod) {
+	sort.Sort(byCompletionIndex(pods))
+	lastIndex := unknownCompletionIndex
+	firstRepeatPos := 0
+	for i, p := range pods {
+		ix := getCompletionIndex(p.Annotations)
+		if ix != lastIndex {
+			rm, left = appendPodsWithSameIndexForRemovalAndRemaining(rm, left, pods[firstRepeatPos:i], lastIndex)
+			firstRepeatPos = i
+			lastIndex = ix
+		}
+	}
+	return appendPodsWithSameIndexForRemovalAndRemaining(rm, left, pods[firstRepeatPos:], lastIndex)
+}
+
+func appendPodsWithSameIndexForRemovalAndRemaining(rm, left, pods []*v1.Pod, ix int) ([]*v1.Pod, []*v1.Pod) {
+	if ix == unknownCompletionIndex {
+		rm = append(rm, pods...)
+		return rm, left
+	}
+	if len(pods) == 1 {
+		left = append(left, pods[0])
+		return rm, left
+	}
+	sort.Sort(controller.ActivePods(pods))
+	rm = append(rm, pods[:len(pods)-1]...)
+	left = append(left, pods[len(pods)-1])
+	return rm, left
+}
+
+func calculatePendingIndexes(pods []*v1.Pod, completions int32) ([]bool, error) {
+	pendingIndexes := make([]bool, completions)
+	for i := range pendingIndexes {
+		pendingIndexes[i] = true
+	}
```

Comment:
return the inverse (nonPending) so we don't need to initialize?

Reply:
I refactored this to directly calculate `firstPendingIndexes` from the list of all pods.